expression/constant_propagation.go

```diff
@@ -290,37 +306,276 @@ func (s *propagateConstantSolver) solve(conditions []Expression) []Expression {
+		rCol, rOk := fun.GetArgs()[1].(*Column)
+		if lOk && rOk {
+			return s.colsFromOuterAndInner(lCol, rCol)
+		}
+	}
+	return nil, nil
+
+}
+
+// deriveConds given `outerCol = innerCol`, derive new expression for specified conditions.
+func (s *propOuterJoinConstSolver) deriveConds(outerCol, innerCol *Column, schema *Schema, fCondsOffset int, visited []bool, fConds bool) []bool {
+	var offset, condsLen int
+	var conds []Expression
+	if fConds {
+		conds = s.fConds
+		offset = fCondsOffset
+		condsLen = len(s.fConds)
+	} else {
+		conds = s.jConds
+		condsLen = fCondsOffset
+	}
+	for k := 0; k < condsLen; k++ {
+		if visited[k+offset] {
+			// condition has been used to retrieve equality relation or contains column beyond children schema.
+			continue
+		}
+		cond := conds[k]
+		if !ExprFromSchema(cond, schema) {
+			visited[k+offset] = true
+			continue
+		}
+		replaced, _, newExpr := tryToReplaceCond(s.ctx, outerCol, innerCol, cond)
+		if replaced {
+			s.jConds = append(s.jConds, newExpr)
+		}
+	}
+	return visited
+}
+
+// propagateColumnEQ propagates expressions like 'outerCol = innerCol' by adding extra filters
+// 'expression(..., innerCol, ...)' derived from 'expression(..., outerCol, ...)' as long as
+// 'expression(..., outerCol, ...)' does not reference columns outside children schemas of join node.
+// Derived new expressions must be appended into join condition, not filter condition.
+func (s *propOuterJoinConstSolver) propagateColumnEQ() {
+	visited := make([]bool, len(s.jConds)+len(s.fConds))
+	s.unionSet = &multiEqualSet{}
+	s.unionSet.init(len(s.columns))
+	var outerCol, innerCol *Column
+	// Only consider column equal condition in joinConds.
+	// If we have column equal in filter condition, the outer join should have been simplified already.
+	for i := range s.jConds {
+		outerCol, innerCol = s.validColEqualCond(s.jConds[i])
+		if outerCol != nil {
+			outerID := s.getColID(outerCol)
+			innerID := s.getColID(innerCol)
+			s.unionSet.addRelation(outerID, innerID)
+			visited[i] = true
+		}
+	}
+	lenJoinConds := len(s.jConds)
+	mergedSchema := MergeSchema(s.outerSchema, s.innerSchema)
+	for i, coli := range s.columns {
+		for j := i + 1; j < len(s.columns); j++ {
+			// unionSet doesn't have iterate(), we use a two layer loop to iterate col_i = col_j relation.
+			if s.unionSet.findRoot(i) != s.unionSet.findRoot(j) {
+				continue
+			}
+			colj := s.columns[j]
+			outerCol, innerCol = s.colsFromOuterAndInner(coli, colj)
+			if outerCol == nil {
+				continue
+			}
+			visited = s.deriveConds(outerCol, innerCol, mergedSchema, lenJoinConds, visited, false)
+			visited = s.deriveConds(outerCol, innerCol, mergedSchema, lenJoinConds, visited, true)
+		}
+	}
+}
+
+func (s *propOuterJoinConstSolver) solve(joinConds, filterConds []Expression) ([]Expression, []Expression) {
+	cols := make([]*Column, 0, len(joinConds)+len(filterConds))
+	for _, cond := range joinConds {
+		s.jConds = append(s.jConds, SplitCNFItems(cond)...)
```

Comment:
If we remove this `SplitCNFItems`, we have to make sure all the join conditions have been split before we come to here? For most cases this assumption should hold, because `planBuilder::buildJoin` would split the `on clause`, but there may exist exceptions, e.g, in `handleCompareSubquery -> buildSemiApply`, I am not sure whether the `condition` has been  split already. Also, this assumption may break for newly added code in future, so I prefer to keep this `SplitCNFItems`.

Reply:
ok